param/__init__.py

```diff
@@ -1201,6 +1203,178 @@ def get_range(self):
+            self._warn('.insert')
+        with self._trigger():
+            super(SelectObjects, self).insert(index, object)
+            self._parameter._objects.insert(index, object)
+
+    def items(self):
+        if self._parameter.names:
+            return self._parameter.names.items()
+        return named_objs(self).items()
+
+    def keys(self):
+        if self._parameter.names:
+            return self._parameter.names.keys()
+        return named_objs(self).keys()
+
+    def pop(self, *args):
+        index = args[0] if args else -1
+        if isinstance(index, int):
+            with self._trigger():
+                super(SelectObjects, self).pop(index)
+                object = self._parameter._objects.pop(index)
+                if self._parameter.names:
+                    self._parameter.names = {
+                        k: v for k, v in self._parameter.names.items()
+                        if v is object
+                    }
+            return
+        if self and not self._parameter.names:
+            raise ValueError(
+                'Cannot pop an object from {clsname}.objects if '
+                'objects was not declared as a dictionary.'
+            )
+        with self._trigger():
+            object = self._parameter.names.pop(*args)
+            super(SelectObjects, self).remove(object)
+            self._parameter._objects.remove(object)
+        return object
+
+    def remove(self, object):
+        with self._trigger():
+            super(SelectObjects, self).remove(object)
+            self._parameter._objects.remove(object)
+            if self._parameter.names:
+                copy = self._parameter.names.copy()
+                self._parameter.names.clear()
+                self._parameter.names.update({
+                    k: v for k, v in copy.items() if v is not object
+                })
+
+    def update(self, objects, **items):
+        if not self._parameter.names:
+            self._parameter.names = named_objs(self)
+        objects = objects.items() if isinstance(objects, dict) else objects
+        with self._trigger():
+            for i, o in enumerate(objects):
+                if not isinstance(o, collections_abc.Sequence):
+                    raise TypeError(
+                        'cannot convert dictionary update sequence element #{i} to a sequence'.format(i=i)
+                    )
+                o = tuple(o)
+                n = len(o)
+                if n != 2:
+                    raise ValueError(
+                        'dictionary update sequence element #{i} has length {n}; 2 is required'.format(i=i, n=n)
+                    )
+                k, v = o
+                self.__setitem__(k, v, trigger=False)
+            for k, v in items.items():
+                self.__setitem__(k, v, trigger=False)
+
+    def values(self):
+        if self._parameter.names:
+            return self._parameter.names.values()
+        return named_objs(self).values()
```

Comment:
I have no way to tell if the above list of methods is exhaustive. It's certainly exhausting! :-) Hope it covers everything!